### PR TITLE
Add freeplay recording with in-app preview and MIDI export

### PR DESCRIPTION
**Pull request by rumycoding**
## Intro
Hi! I’ve been using Neothesia and really enjoy it. To support personal practice, I added a freeplay recording feature that captures live MIDI input, converts the take into a playable song, and allows saving it locally as a MIDI file.

## Summary
This PR introduces an in-app recording flow: when recording stops, the app builds a preview of the take directly—no external reload required. It also adds a preview UI with record/save/play/pause, seeking, progress tracking, measure markers, and waterfall/note rendering, all integrated with existing scene components. Supporting changes include in-memory SMF construction and safeguards around playback math to handle zero-length or very short recordings gracefully.

## File changed
- [neothesia/src/scene/freeplay/mod.rs]: adds freeplay recording, in-app preview playback, progress/seek UI, and MIDI export.
- [midi-file/src/file.rs]: adds in-memory MIDI loading so recorded takes can be previewed without writing a temporary file.
- [neothesia/src/scene/playing_scene/midi_player.rs]: supports configurable lead-in timing and safer seek/progress handling for preview playback.
- [midi-file/src/playback.rs]: guards playback progress calculations when a recording has zero length.
- [neothesia/src/scene/playing_scene/keyboard.rs]: lets freeplay preview apply generated song configuration to the keyboard UI.
- [neothesia/src/scene/playing_scene/mod.rs]: exposes the MIDI player module for reuse in freeplay preview.

### CODE REVIEW

File: neothesia/src/scene/playing_scene/midi_player.rs [hunk 1]
```diff
@@ -154,11 +167,21 @@ impl MidiPlayer {
     }
 
     pub fn percentage_to_time(&self, p: f32) -> Duration {
-        Duration::from_secs_f32((p * self.playback.length().as_secs_f32()).max(0.0))
+        let length = self.playback.length();
+        if length.is_zero() {
+            Duration::ZERO
+        } else {
+            Duration::from_secs_f32((p * length.as_secs_f32()).max(0.0))
+        }
```
Comment (PolyMeilex) [inline neothesia/src/scene/playing_scene/midi_player.rs]: This change does nothing as far as I can tell. Muliplying by 0 already yelds zero.

File: midi-file/src/playback.rs [hunk 2]
```diff
@@ -94,7 +94,12 @@ impl PlaybackState {
     }
 
     pub fn percentage(&self) -> f32 {
-        self.running.as_secs_f32() / self.length().as_secs_f32()
+        let length = self.length();
+        if length.is_zero() {
+            0.0
```
Comment (PolyMeilex) [inline midi-file/src/playback.rs]: This branch looks unreachable

File: neothesia/src/scene/playing_scene/midi_player.rs [hunk 3]
```diff
@@ -154,11 +167,21 @@ impl MidiPlayer {
...
     }
 
     pub fn time_to_percentage(&self, time: &Duration) -> f32 {
-        time.as_secs_f32() / self.playback.length().as_secs_f32()
+        let length = self.playback.length();
+        if length.is_zero() {
+            0.0
```
Comment (PolyMeilex) [inline neothesia/src/scene/playing_scene/midi_player.rs]: This branch looks unreachable

File: neothesia/src/scene/freeplay/mod.rs [hunk 4]
```diff
@@ -32,9 +45,193 @@ pub struct FreeplayScene {
     nuon: nuon::Ui,
     mouse_to_midi_state: MouseToMidiEventState,
     deduced_chord_name: String,
+    recorder: FreeplayRecorder,
+    recorder_status: String,
+    preview_player: Option<MidiPlayer>,
+    preview_waterfall: Option<WaterfallRenderer>,
+    preview_note_labels: Option<NoteLabels>,
```
Comment (PolyMeilex) [inline neothesia/src/scene/freeplay/mod.rs]: Those fields should be folded into a new `Option<T>` field, 
aka, `clear_preview()`  should be able to just do one `self.preview_state = None`  instead of 3 separate clears.

File: neothesia/src/scene/freeplay/mod.rs [hunk 5]
```diff
@@ -32,9 +45,193 @@ pub struct FreeplayScene {
...
+    timestamp: Duration,
+    channel: u8,
+    message: MidiMessage,
+}
+
+struct FreeplayRecorder {
+    started_at: Option<Instant>,
+    recorded_duration: Duration,
+    events: Vec<RecordedMidiEvent>,
+    active_notes: HashSet<(u8, u8)>,
```
Comment (PolyMeilex) [inline neothesia/src/scene/freeplay/mod.rs]: Similarly to `FreeplayScene` struct comment, instead of manually clearing fields, and using `started_at` as an indicator of the current state of the recorder, the fields should be moved into a separate `in_progress: Option<RecordingInProgressState>` field (or something along those lines).

Starting the recording should create a new state and set is as `in_progress: Some()`
Stoping the recoding should just clear that one field to None. So it's never posible to leave some leftover data behing by accident.

**Review by PolyMeilex — CHANGES_REQUESTED** [review]
btw code forrmating is quite random, folow the official Rust formating rules.

**Comment by PolyMeilex** [comment]
Looks like a good idea btw!

File: neothesia/src/scene/freeplay/mod.rs (hunk 4 above)
Comment (rumycoding) [reply neothesia/src/scene/freeplay/mod.rs]: grouped into PreviewState

File: neothesia/src/scene/freeplay/mod.rs (hunk 5 above)
Comment (rumycoding) [reply neothesia/src/scene/freeplay/mod.rs]: grouped this into
```
enum RecorderState {
    Idle,
    Recording(RecordingInProgressState),
    Recorded(RecordedTake),
}
```

**Comment by rumycoding** [comment]
hi @PolyMeilex, Thanks for the detailed review. I’ve done my best to revise the structure and style so the change is easier to follow and reads more cleanly. While working through the feedback, I also found and fixed a potential bug in keyboard.rs: some MIDI devices encode note release as NoteOn with velocity 0, so that case now gets handled explicitly.

If you see anything else that should be adjusted, please leave a comment!😄

**Comment by PolyMeilex** [comment]
> I also found and fixed a potential bug in keyboard.rs: some MIDI devices encode note release as NoteOn with velocity 0, so that case now gets handled explicitly.

I already normalize the midi input long before it reaches the app code to avoid the need for doing if vel == 0 checks in multiple places, it's done right here https://github.com/PolyMeilex/Neothesia/blob/master/neothesia%2Fsrc%2Finput_manager%2Fmod.rs#L37


I will take a look at the code tomorrow, but at a glance it looks like what I had in mind so should be good.

**Comment by rumycoding** [comment]
ok, have reverted that part. Thanks~

**Comment by rumycoding** [comment]
Also, added test coverage for edge cases where the recording contains only pedal or note-off MIDI events, since those should not generate a preview.

**Review by PolyMeilex — APPROVED** [review]

**Comment by PolyMeilex** [comment]
Thanks!
